hacspec/src/math_util.rs

```diff
@@ -385,3 +387,291 @@ pub fn ct_div<T: Integer + Copy>(a: T, d: T) -> (T, T) {
+        }
+        for j in 0..b.len(){
+            if b[j] == 0 {
+                continue;
+            }
+
+            if i + j > a.len() - 2 {
+                // modulo irr
+                // factor is the coeff
+                let factor = a[i] * b[j];
+                result[1+((i + j) % (a.len() -1))] = result[1 + (i + j) % (a.len() - 1)] -factor * irr[1];
+                result[(i + j) % (a.len() -1)] = result[(i + j) % (a.len() - 1)] -factor * irr[0];
+                continue;
+            }
+            result[i + j] = result[i + j] + a[i] * b[j];
+        }
+    }
+    if modulo > 0 {
+        for i in 0..result.len(){
+            result[i] = result[i] % modulo;
+
+        }
+    }
+    make_positive(&result,modulo)
+}
+
+pub fn mul_poly_native(a:&Seq<i128>,b:&Seq<i128>,modulo:i128) -> Seq<i128>{
+    let mut out:Seq<i128> = Seq::new(a.len()+b.len());
+    for i in 0..a.len() {
+        if a[i] == 0{
+            continue;
+        }
+        for j in 0..b.len() {
+            out[i + j] = (a[i] * b[j] + out[i + j]) % modulo;
+        }
+    }
+    make_positive(&out,modulo)
+}
+
+
+
+///simple poly multiplication (schoolbook with sparse) O(n²)
+pub fn mul_poly(a:&Seq<i128>,b:&Seq<i128>,modulo:i128) -> Seq<i128>{
+    assert!(a.len()==b.len(),true);
+    let mut result:Seq<i128> = Seq::new(a.len());
+    for i in 0..a.len(){
+        if a[i] == 0{
+            continue;
+        }
+        for j in 0..b.len(){
+            if b[j] == 0{
+                continue;
+            }
+            if i + j > a.len() -1 {
+                panic!("Overflow");
+            }
+            result[i + j] = (result[i + j] + (a[i] * b[j])) % modulo;
+        }
+    }
+    result
+}
+fn monomial_poly(poly:&Seq<i128>,size:usize) ->Seq<i128>{
+    let mut result:Seq<i128> = Seq::new(size);
+    for i in 0..poly.len(){
+        result[i] = poly[i];
+    }
+    result
+}
+
+fn is_empty(poly:&Seq<i128>)->bool{
+    let mut result = true;
+    for i in 0..poly.len(){
+        if poly[i] != 0{
+            result = false;
+            break;
+        }
+    }
+    result
+}
+
+pub fn deg(poly:&Seq<i128>) -> usize{
+    let mut deg = 0;
+    for i in 0..poly.len()-1{
+        if poly[poly.len() - 1 - i] != 0{
+            deg = poly.len() - 1 -i;
+            break;
+        }
+    }
+    deg
+}
+
+pub fn weight(poly:&Seq<i128>)->usize{
+    let tmp = Seq::from_seq(poly);
+    let mut weight = 0;
+    for i in 0..tmp.len(){
+        if tmp[i] != 0{
+            weight = weight + 1;
+        }
+    }
+    weight
+}
+
+
+pub fn leading_coef(poly:&Seq<i128>) -> i128{
+    poly[deg(poly)]
+}
+
+pub fn make_positive(poly:&Seq<i128>, q:i128)-> Seq<i128>{
+    let mut result = Seq::from_seq(poly);
+    for i in 0..poly.len(){
+        if result[i] < 0 {
+            result[i] = poly[i] + q;
+        }
+    }
+    result
+}
+
+
+pub fn round(poly:&Seq<i128>, round_to:i128, q:i128)->Seq<i128>{
+    // viewed as integers between−(q−1)/2 and(q−1)/2
```

Comment:
So this is a very specific function with a very generic name 😄 
Can you give a better name and add documentation what exactly it is doing?

Reply:
Sure :smile: